src/scene/renderer/forward-renderer.js

```diff
@@ -1325,26 +1314,33 @@ class ForwardRenderer {
 
                 if (material !== prevMaterial) {
                     this._materialSwitches++;
+                    material._scene = scene;
 
                     if (material.dirty) {
                         material.updateUniforms(device, scene);
                         material.dirty = false;
                     }
 
+                    // if material has dirtyBlend set, notify scene here
+                    if (material._dirtyBlend) {
+                        scene.layers._dirtyBlend = true;
+                    }
+
                     if (!drawCall._shader[pass] || drawCall._shaderDefs !== objDefs || drawCall._lightHash !== lightHash) {
                         if (!drawCall.isStatic) {
                             const variantKey = pass + '_' + objDefs + '_' + lightHash;
                             drawCall._shader[pass] = material.variants[variantKey];
                             if (!drawCall._shader[pass]) {
-                                this.updateShader(drawCall, objDefs, null, pass, sortedLights);
+                                drawCall.updatePassShader(scene, pass, null, sortedLights);
                                 material.variants[variantKey] = drawCall._shader[pass];
                             }
                         } else {
-                            this.updateShader(drawCall, objDefs, drawCall._staticLightList, pass, sortedLights);
+                            drawCall.updatePassShader(scene, pass, drawCall._staticLightList, sortedLights);
                         }
```

Comment:
Non-clarity of this blocks also still bothers be, and I'll refactor it later, probably soon as I might need it. I don't like that the caller of updateShader needs to calculate the variant key, and add the shader to material.variants. This should be completely internal ideally, the key computed from shader options generated, and not on some very few variables. I suspect that at the moment, this does not allow us to have all the variations we need, as the keys are the same for them.